# Artificial Neural Network

### Importing the libraries

In [74]:
#!pip install numpy
#!pip install pandas
#!pip install matplotlib
#!pip install tensorflow
#!pip install sklearn

In [75]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf

In [76]:
tf.__version__

'2.10.0'

## Part 1 - Data Preprocessing

### Importing the dataset

In [77]:
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1]
y = dataset.iloc[:, -1]

In [78]:
print(X)

      CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619   Senegal  Female   42       2       0.00              1   
1             608    Guinea  Female   41       1   83807.86              1   
2             502   Senegal  Female   42       8  159660.80              3   
3             699   Senegal  Female   39       1       0.00              2   
4             850    Guinea  Female   43       2  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9995          771   Senegal    Male   39       5       0.00              2   
9996          516   Senegal    Male   35      10   57369.61              1   
9997          709   Senegal  Female   36       7       0.00              1   
9998          772    Canada    Male   42       3   75075.31              2   
9999          792   Senegal  Female   28       4  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  
0           

In [79]:
print(y)

0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64


### Encoding categorical data

Label Encoding the "Gender" column

In [80]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
#X[:, 2] = le.fit_transform(X[:, 2])
X.loc[:, 'Gender'] = le.fit_transform(X.loc[:, 'Gender'])

C:\Users\balde\AppData\Local\Temp\ipykernel_7604\2589284813.py:4: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  X.loc[:, 'Gender'] = le.fit_transform(X.loc[:, 'Gender'])


In [81]:
print(X)

      CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619   Senegal       0   42       2       0.00              1   
1             608    Guinea       0   41       1   83807.86              1   
2             502   Senegal       0   42       8  159660.80              3   
3             699   Senegal       0   39       1       0.00              2   
4             850    Guinea       0   43       2  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9995          771   Senegal       1   39       5       0.00              2   
9996          516   Senegal       1   35      10   57369.61              1   
9997          709   Senegal       0   36       7       0.00              1   
9998          772    Canada       1   42       3   75075.31              2   
9999          792   Senegal       0   28       4  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  
0           

One Hot Encoding the "Geography" column

In [82]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = ct.fit_transform(X)

In [83]:
print(X)

[[0.0000000e+00 0.0000000e+00 1.0000000e+00 ... 1.0000000e+00
  1.0000000e+00 1.0134888e+05]
 [0.0000000e+00 1.0000000e+00 0.0000000e+00 ... 0.0000000e+00
  1.0000000e+00 1.1254258e+05]
 [0.0000000e+00 0.0000000e+00 1.0000000e+00 ... 1.0000000e+00
  0.0000000e+00 1.1393157e+05]
 ...
 [0.0000000e+00 0.0000000e+00 1.0000000e+00 ... 0.0000000e+00
  1.0000000e+00 4.2085580e+04]
 [1.0000000e+00 0.0000000e+00 0.0000000e+00 ... 1.0000000e+00
  0.0000000e+00 9.2888520e+04]
 [0.0000000e+00 0.0000000e+00 1.0000000e+00 ... 1.0000000e+00
  0.0000000e+00 3.8190780e+04]]


### Splitting the dataset into the Training set and Test set

In [84]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### Feature Scaling

**Note**: Scaling is a critical step to Neural Networks, so we always scale!

In [85]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Part 2 - Building the ANN

### Initializing the ANN

Sequential Layer

In [86]:
ann = tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

**Note**: More commonly, for the activation function we'll use:

*   Rectifier function for the hidden layer
*   Sigmoid function for the output layer



In [87]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the second hidden layer

In [88]:
# "6 Nodes" has resulted in the best accuracy so far
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the output layer

I have used "sigmoid" as it is a binary classification.
For multiclass, use "softmax" instead.

In [89]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

## Part 3 - Training the ANN

### Compiling the ANN

Understanding the parameters:

*   optimizer: updates the weights for better accuracy
*   loss: the loss function
  * binary_crossentropy: Binary classification:
  * categorical_crossentropy: Multiclass classification

In [90]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

### Training the ANN on the Training set

In [91]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 [==============================] - 1s 1ms/step - loss: 0.6756 - accuracy: 0.6140
Epoch 2/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4928 - accuracy: 0.7969
Epoch 3/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4495 - accuracy: 0.7979
Epoch 4/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4356 - accuracy: 0.8015
Epoch 5/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4286 - accuracy: 0.8074
Epoch 6/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4240 - accuracy: 0.8115
Epoch 7/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4205 - accuracy: 0.8145
Epoch 8/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4166 - accuracy: 0.8164
Epoch 9/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4111 - accuracy: 0.8200
Epoch 10/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4032 - accura

## Part 4 - Making the predictions and evaluating the model

### Predicting the result of a single observation

Let's use our ANN model to predict if the customer with the following informations will leave the bank: 

Geography: Senegal

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

In [92]:
# Checking the position of the dummy variable
X[0]

array([0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 6.1900000e+02,
       0.0000000e+00, 4.2000000e+01, 2.0000000e+00, 0.0000000e+00,
       1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0134888e+05])

In [93]:
dfX = pd.DataFrame(X, columns=['Senegal', 'Guinea', 'Canada',  'CreditScore', 'Gender',	'Age',	'Tenure',	'Balance',	'NumOfProducts',	'HasCrCard',	'IsActiveMember',	'EstimatedSalary'])
dfy = pd.DataFrame(y)
# concat by colmun (axis=1)
df = pd.concat([dfX, dfy], axis=1)

In [94]:
df

Senegal  Guinea  Canada  CreditScore  Gender   Age  Tenure    Balance  \
0         0.0     0.0     1.0        619.0     0.0  42.0     2.0       0.00   
1         0.0     1.0     0.0        608.0     0.0  41.0     1.0   83807.86   
2         0.0     0.0     1.0        502.0     0.0  42.0     8.0  159660.80   
3         0.0     0.0     1.0        699.0     0.0  39.0     1.0       0.00   
4         0.0     1.0     0.0        850.0     0.0  43.0     2.0  125510.82   
...       ...     ...     ...          ...     ...   ...     ...        ...   
9995      0.0     0.0     1.0        771.0     1.0  39.0     5.0       0.00   
9996      0.0     0.0     1.0        516.0     1.0  35.0    10.0   57369.61   
9997      0.0     0.0     1.0        709.0     0.0  36.0     7.0       0.00   
9998      1.0     0.0     0.0        772.0     1.0  42.0     3.0   75075.31   
9999      0.0     0.0     1.0        792.0     0.0  28.0     4.0  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0               1.0        1.0             1.0        101348.88       1  
1               1.0        0.0             1.0        112542.58       0  
2               3.0        1.0             0.0        113931.57       1  
3               2.0        0.0             0.0         93826.63       0  
4               1.0        1.0             1.0         79084.10       0  
...             ...        ...             ...              ...     ...  
9995            2.0        1.0             0.0         96270.64       0  
9996            1.0        1.0             1.0        101699.77       0  
9997            1.0        0.0             1.0         42085.58       1  
9998            2.0        1.0             0.0         92888.52       1  
9999            1.0        1.0             0.0         38190.78       0  

[10000 rows x 13 columns]

In [95]:
# Since we scaled the training set when training the algorithm, 
# we also have to scale the new test data
customer = [[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]
customer_scaled = sc.transform(customer)
y_pred = ann.predict(customer_scaled)

1/1 [==============================] - 0s 101ms/step


In [96]:
print(y_pred)

[[0.03163132]]


**Conclusion**

There is less than 4% chances the customer will leave to bank.

Therefore, our ANN model predicts that this customer stays in the bank!

**Note:** The "predict" method always expects a 2D array as the format of its inputs hence the customer double brackets.

### Predicting the Test set results

In [97]:
print(y_pred.reshape(1, -1))

[[0.03163132]]


In [98]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)

y_test_array = y_test.values

# reshaping on n rows - 1 column
y_pred_reshaped = y_pred.reshape(len(y_pred), 1)
y_test_reshaped = y_test_array.reshape(len(y_test),1)
y_pred_test = np.concatenate((y_pred_reshaped, y_test_reshaped), 1)
print(y_pred_test)

63/63 [==============================] - 0s 847us/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


### Making the Confusion Matrix

In [99]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1517   78]
 [ 207  198]]


0.8575

## Conclusion

Our ANN model has a pretty good accuracy of 86%.
Would you find a better one, please email me.

This project was a fun one! But it's also important as it helps improve the business by keeping more customers.
I am looking forward to take a more challenging one.
